sw/host/opentitanlib/src/io/emu.rs

```diff
@@ -0,0 +1,83 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::transport::Result;
+use serde::{Deserialize, Serialize};
+use std::collections::HashMap;
+use std::time::Duration;
+use thiserror::Error;
+
+/// Error related to the `Emulator` trait. These error messages will be printed in the context of
+/// a TransportError::EmuError, that is "Emulator error: {}".  So including the words "error" or
+/// "Emulator" in texts below will probably be redundant.
+#[derive(Error, Debug, Serialize, Deserialize)]
+pub enum EmuError {
+    #[error("Invalid argument name {0}")]
+    InvalidArgumetName(String),
+    #[error("Argument {0} has invalid value {1}")]
+    InvalidArgumentValue(String, String),
+    #[error("Start failed with cause {0}")]
+    StartFailureCause(String),
+    #[error("Stop failed with cause {0}")]
+    StopFailureCause(String),
+    #[error("Timeout during performing command: {0} timeout: {1}")]
+    Timeout(String, u64),
+    #[error("Can't restore resource to initial state: {0}")]
+    ResetError(String),
+    #[error("Runtime error {0}")]
+    RuntimeError(String),
+}
+
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub enum EmuValue {
+    Empty,
+    String(String),
+    FilePath(String),
+    StringList(Vec<String>),
+    FilePathList(Vec<String>),
+}
+
+#[derive(PartialEq, Eq, Clone, Copy, Debug, Serialize, Deserialize)]
+pub enum EmuState {
+    Off,
+    On,
+    Busy,
+    Error,
+}
+
+/// A trait which represents a `Emulator` instance
+pub trait Emulator {
+    /// Acquires state of emulator
+    fn get_state(&self) -> Result<EmuState>;
+
+    /// Waits for the specified amount of time passed in `timeout`, for first state transition
+    /// of the Emulator form `Busy` to one of states defined in `EmuState` other that `Busy`.
+    /// If the transition from `Busy` to another state does not take place until
+    /// the `timeout` expires, the function will return an `Timeout` error.
+    /// In case of any transition of Emulator to `Error` state,
+    /// the function immediately returns `RuntimeError` with description of a problem.
+    fn wait_until_busy(&self, timeout: Duration) -> Result<()>;
+
+    /// Start Emulator with provided arguments. Parameter `timeout`
```

Comment:
Oh, does this function only return after emulation has started (with timeout)? Or does it return immediately?  In the latter case, what exactly is the timeout for?  In the former case, would it make sense to specify that Err() will be returned if the state ends in anything other than On?

Reply:
Michał and I discussed this, and agreed to leaving out the `wait()` function until such time when we have a better idea how to handle asynchronous method calls in the session process.  So the trait will only have `start()`, `stop()` and `get_state()` for now.